Fix segfault (null deref) in named initialisation of nested anonymous union

el-remph opened the pull request:
This seems very specific but I did happen upon it organically. What were the chances? lol

Also adds a regression test, which causes previous chibicc version to deref NULL and segfault (the cc1 process dies with SIGSEGV, which causes the driver to `exit(1)`).

I think I'm right to assume that the only way `mem->name` can be NULL is if `mem` is an anonymous union/struct? Just wanted to make sure

fuhsnn commented:
`mem->name` will be null for unnamed bitfields too, for example this will trigger the newly added assert:
```
struct { int : 7; int foo; } s = {.foo = 0};
```
IMO unnamed bitfields should be filtered out earlier though.

el-remph committed:
Fix segfault (null deref) in named initialisation of nested anonymous unions

el-remph commented:
@fuhsnn Thanks, I hadn't thought of that. Added that to the test, and fixed the code to allow it, and still prevent `mem->name->len` from being reached if `mem->name`  is NULL
